pcdscalc/be_lens_calcs.py

```diff
@@ -937,3 +930,395 @@ def find_z_pos(energy, lens_set, spot_size_fwhm, material=None,
+        distances < np.max(z_offset + np.asarray(z_range)))
+
+    sets = sets[good_sets]
+    size_range_min = np.asarray([
+            calc_beam_fwhm(
+                energy,
+                list(chain(*list(zip(set, t_lens_radii)))),
+                distance=z_offset - min(z_range),
+                fwhm_unfocused=beam_size_unfocused,
+                printsummary=False,
+            ) for set in sets])
+
+    size_range_max = np.asarray([
+        calc_beam_fwhm(
+            energy,
+            list(chain(*list(zip(set, t_lens_radii)))),
+            distance=z_offset - max(z_range),
+            fwhm_unfocused=beam_size_unfocused,
+            printsummary=False,
+        ) for set in sets])
+
+    sizes = sizes[good_sets]
+    effrads = effrads[good_sets]
+    distances = distances[good_sets]
+    foclens = foc_lens[good_sets]
+    transms = np.asarray(
+        [lens_transmission(r=ter, fwhm=beam_size_unfocused, energy=energy)
+            for ter in effrads])
+    Nlenses_s = np.sum(sets, 1)
+
+    # used for testing
+    num = []
+    f_m = []
+    min_um = []
+    max_um = []
+    t_percent = []
+
+    resstring = " N   f/m   Min/um   Max/um   T/%  Set  \n"
+    zips = list(
+        zip(sets, size_range_min, size_range_max,
+            effrads, transms, Nlenses_s, foclens)
+    )
+
+    t = "{:2d} {:5.2f} {:8.1f} {:8.1f} {:5.1f}  "
+    for n, z in enumerate(zips):
+        the_set, sizemin, sizemax, eff_rad, transm, n_lenses, foclen = z
+        logger.debug('eff_rad: %s', eff_rad)
+        logger.debug('n_lenses: %s', n_lenses)
+
+        num.append(n)
+        f_m.append(foclen)
+        min_um.append(sizemin * 1e6)
+        max_um.append(sizemax * 1e6)
+        t_percent.append(transm * 100)
+
+        resstring += t.format(n, foclen, sizemin * 1e6,
+                              sizemax * 1e6, transm * 100)
+        resstring += ", ".join(
+            [
+                "%d x %dum" % (setLensno, t_lens_radii[m] * 1e6)
+                for m, setLensno in enumerate(the_set)
+                if setLensno > 0
+            ]
+        )
+        resstring += "\n"
+    logger.info('\n %s', resstring)
+    return num, f_m, min_um, max_um, t_percent
+
+
+def lens_transmission(r, fwhm, n=1, energy=None, id_material="IF1",
+                      lens_thicknes=None):
+    """
+    Find the CRL (Compound Refractive Lens) transmission.
+
+    Parameters
+    ----------
+    radius : float
+        Effective radius of curvature
+    fwhm : float
+        Incident beam size on the lens in meters
+    num : int
+        Number of lenses in the stack
+    energy : number
+        Photon energy
+    id_material : str
+        Lens material
+
+    Returns
+    -------
+    trans : float
+    """
+    lens_thicknes = lens_thicknes or APEX_DISTANCE
+    id_material = check_id(id_material)
+    # TODO: what should I do with the energy here
+    # energy = getE(energy=energy, correctEv=True)
```

Comment:
this function requires a PV, should we have that in the calculations or is this up to to user to pass in the correct energy when using the classes?
https://github.com/pcdshub/old-hutch-configs/blob/18c7769d66a647ab558ad7ebc34d019fc98fbf48/cxi/blutil/calc.py#L10-L40

Reply:
There shouldn't be anything that requires a PV in this codebase - values only.

Reply:
The previous codebase checked a PV if you didn't provide an energy value. I'm not sure exactly how/where to implement this behavior... perhaps not until we get to `pcdsdevices` and you need to move to a specific beam size?

Generally when you're planning an experiment the current x-ray energy does not match what it will be during your beam time.

Reply:
@ZLLentz what's the issue in requiring the user to be explicit about the energy as an input to these functions? Or rather, why do you think that functionality from the old codebase need to be included here?

Reply:
When you call these functions you should call them with the energy you expect to have during the beam time. There may be some cases during a beam time when you want to call a function with "current beam energy". These are mostly localized around "I want to move my calculation motor without repeatedly updating the beam energy parameter"

Reply:
I see - perhaps our LCLS object should have an easily toggleable subscription that calls the new equivalent of `setE` for setting the default energy when none is supplied by kwarg?

Reply:
I agree Ken- the LCLS or BeamStats objects should have an easy hook for this, and then devices that need to do that calculation should set it up.

Reply:
Ok sounds good, I'll do this in the other PR (https://github.com/pcdshub/pcdsdevices/pull/633) probably because I might need it in the LensStack class, but maybe not yet... regardless it will be good to add it - thank you

Reply:
> I see - perhaps our LCLS object should have an easily toggleable subscription that calls the new equivalent of `setE` for setting the default energy when none is supplied by kwarg?

@klauer  when you get a chance can you please elaborate on this^?

Will it look similar to this: (in the LCSL object)
```
photon_ev_hxr = Cpt(EpicsSignalRO, 'SIOC:SYS0:ML00:AO627', kind='hinted',
                        doc='Photon eV HXR [eV]')

    @photon_ev_hxr.sub_value
    def _energy_changed(self, *args, value, ** kwargs):
        self.set_energy(value)
        print(f'----VALUE CHANGED -------{value}')

    def set_energy(self, value):
        global _photon_energy
        _photon_energy = value

    def get_energy(self):
        return _photon_energy
```
And then in the LensStack object I would get this value from the LCLS object?
```
lcls = LCLS()
lcls.get_energy()
```

Reply:
I was thinking something along the lines of:

```python
photon_ev_hxr.subscribe(_set_current_energy)

def get_energy():
   if user_specified_energy is not None:
       return user_specified_energy
    return _current_energy

def other_calculation_function(, ..., energy=None):
    if energy is None:
        energy = get_energy()
```

It's a lot of customization of one parameter - but it's apparently an important one to be flexible with here.

Reply:
I'm still a bit confused here - i'm so sorry.
Here is what i understand:
The `photon_ev_hxr.subscribe(_set_current_energy)` is in the `beam_stats` `LCLS` object. Here is where I will have access to the `current_energy` - but I don't need to know here about the `user_specified_energy` - or do i? - i'm not doing any calculations here. 
I would need to know about the user specified energy in the `LensStack` and in the calculations... 

I need to think about this a bit more, i might be overcomplicating things here

Reply:
The `user_specified_energy` is the user-provided energy that acts as an override for the current energy provided by the LCLS subscription callback.

Maybe I'm missing something here, too - not sure where the confusion is coming in. (@ZLLentz?)

Reply:
> I like (3) - let's get this rolling and in the hands of those that can use it 👍

Sounds good, at least if people start using it we'll get some feedback (maybe) from them and hopefully some of the questions/concerns will be answered...